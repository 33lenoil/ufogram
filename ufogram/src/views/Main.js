import React, { useState } from 'react';
import Activity from './Activity';
import { useNavigate } from 'react-router-dom';
import { useLocation} from 'react-router-dom';
import getUser from '../api/getUser';

export default function Main() {

    const navigate = useNavigate();

    const location = useLocation();

    const [usernameToSearch, setUsernameToSearch] = useState('');

    const handleMyProfile = () => {
        navigate('/userprofile', { state: { userId: location.state.userId, username: location.state.username, self: true, sName:  location.state.username, sId: location.state.userId, users: location.state.users, followed: true } });
    };

    const handleCreateNewPost = () => {
        console.log('Create New Post Handler Called');
        navigate('/newpost', { state: { userId: location.state.userId, username: location.state.username, users: location.state.users } });
    };

    const handleSearchUserName = (usernameEvent) => {
        setUsernameToSearch(usernameEvent.target.value);
    }

    const handleSearchUser = async () => {
        for (let i = 0; i < location.state.users.length; i++) {
            if (location.state.users[i].username === usernameToSearch) {
                console.log(location.state.users[i].id);
                const searchResponse = await getUser(location.state.users[i].id);
                for (let j = 0; j < searchResponse.data.followers.length; j++) {
                    if (searchResponse.data.followers[j].id === location.state.userId) {
                        navigate('/userprofile', { state: { userId: location.state.userId, username: location.state.username, self: location.state.username ===  usernameToSearch, sName: usernameToSearch, sId: location.state.users[i].id, users: location.state.users, followed: true } });
                        return;
                    }
                }
                navigate('/userprofile', { state: { userId: location.state.userId, username: location.state.username, self: location.state.username ===  usernameToSearch, sName: usernameToSearch, sId: location.state.users[i].id, users: location.state.users, followed: false } });
                return;
            }
        }
    }


    return (
        <div style={{display: "flex", justifyContent: "space-evenly"}}>
            <div style={{display: "flex", justifyContent: "center", position: "fixed", width: "100%", background: "#8769b6"}}>
                <h1>UFOgram { location.state.username }</h1>
                <div style={{ display: "flex", flexDirection: "row" }}>
                <input type="text" name="SearchUser" onChange={ handleSearchUserName }/>
                <button type="button" title="searchUser" onClick={ handleSearchUser }>Search</ button>
                </div>
                <div>
                    <button type="button" title="My Profile" onClick={handleMyProfile}>My Profile</ button>
                    <button type="button" title="Create New Post" onClick={handleCreateNewPost}>Create New Post</ button>
                </div>
            </div>
            <div style={{display: "flex", width: "100%", justifyContent: "center", marginTop: "80px"}}>
                <Activity />
            </div >
        </div>
    )
}