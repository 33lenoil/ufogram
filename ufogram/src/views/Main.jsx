import React, { useState } from 'react';
import { useNavigate, useLocation } from 'react-router-dom';
import Activity from './Activity';
import getUser from '../api/getUser';

export default function Main() {
  const navigate = useNavigate();

  const location = useLocation();

  const [usernameToSearch, setUsernameToSearch] = useState('');

  const handleMyProfile = () => {
    navigate('/userprofile', {
      state: {
        userId: location.state.userId,
        username: location.state.username,
        self: true,
        sName: location.state.username,
        sId: location.state.userId,
        users: location.state.users,
        followed: true,
      },
    });
  };

  const handleCreateNewPost = () => {
    // console.log('Create New Post Handler Called');
    navigate('/newpost', { state: { userId: location.state.userId, username: location.state.username, users: location.state.users } });
  };

  const handleSearchUserName = (usernameEvent) => {
    setUsernameToSearch(usernameEvent.target.value);
  };

  const handleSearchUser = async () => {
    const userToSearch = location.state.users.find((user) => user.username === usernameToSearch);

    if (userToSearch) {
      const searchResponse = await getUser(userToSearch.id);
      const isFollowed = searchResponse.data.followers.some(
        (follower) => follower.id === location.state.userId,
      );

      const userProfileState = {
        userId: location.state.userId,
        username: location.state.username,
        self: location.state.username === usernameToSearch,
        sName: usernameToSearch,
        sId: userToSearch.id,
        users: location.state.users,
        followed: isFollowed,
      };

      navigate('/userprofile', {
        state: userProfileState,
      });
    } else {
      // Handle the case where the user is not found
    }
  };

  return (
    <div style={{ display: 'flex', justifyContent: 'space-evenly' }}>
      <div style={{
        display: 'flex', justifyContent: 'center', position: 'fixed', width: '100%', background: '#8769b6',
      }}
      >
        <h1>
          UFOgram
          {' '}
          { location.state.username }
        </h1>
        <div style={{ display: 'flex', flexDirection: 'row' }}>
          <input type="text" name="SearchUser" onChange={handleSearchUserName} />
          <button type="button" title="searchUser" onClick={handleSearchUser}>Search</button>
        </div>
        <div>
          <button type="button" title="My Profile" onClick={handleMyProfile}>My Profile</button>
          <button type="button" title="Create New Post" onClick={handleCreateNewPost}>Create New Post</button>
        </div>
      </div>
      <div style={{
        display: 'flex', width: '100%', justifyContent: 'center', marginTop: '80px',
      }}
      >
        <Activity userId={location.state.userId} selfKind={0} state={useLocation.state} />
      </div>
    </div>
  );
}
