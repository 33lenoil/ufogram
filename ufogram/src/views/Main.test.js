/**
 * @jest-environment jsdom
 */

import React, { useState } from 'react';
import '@testing-library/jest-dom';
import { render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { BrowserRouter as Router, Routes, Route } from 'react-router-dom';
import userEvent from '@testing-library/user-event';
import Main from './Main';

jest.mock("react-router-dom", () => ({
    ...jest.requireActual("react-router-dom"),
    useLocation: () => ({
      pathname: "localhost:3001/main",
      state: { userId: 1, username: 'lionelhu', users: [] }
    })
  }));


test('renders title', () => {
    render(
      <Router>
          {/* <Routes>
              <Route path="/signup" element={<Login />} />
          </Routes> */}
          <Main />
      </Router>
    );
    const linkElement = screen.getByText(/UFOgram/);
    expect(linkElement).toBeInTheDocument();
  });

test('renders my profile button', () => {
    render(
        <Router>
            {/* <Routes>
                <Route path="/signup" element={<Login />} />
            </Routes> */}
            <Main />
        </Router>
      );
  const linkElement = screen.getByRole('button', {
    name: /My Profile/
  })
  expect(linkElement).toBeInTheDocument();
});

test('renders new post button', () => {
    render(
        <Router>
            {/* <Routes>
                <Route path="/signup" element={<Login />} />
            </Routes> */}
            <Main />
        </Router>
      );
  const linkElement = screen.getByRole('button', {
    name: /Create New Post/
  })
  expect(linkElement).toBeInTheDocument();
});

/**
 * Snapshot Testing
 */

test('the component matches the snapshot', () => {
  const component = renderer.create(<Router>
    {/* <Routes>
        <Route path="/signup" element={<Signup />} />
    </Routes> */}
    <Main />
</Router>);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

/**
 *  Testing
 */

// test('welcome is displayed after click the registration button', async () => {
//   render(<Login />);
//   const buttonElement = screen.getByRole('button');
//   await userEvent.click(buttonElement);
//   const welcomeElement = screen.getAllByLabelText(/welcome/i);
//   expect(welcomeElement).toBeInTheDocument();
// });

// test('Username is displayed after clicking on registration button', async () => {
//   render(<Login />);
//   const usernameElement = screen.getByTestId('uname');
//   await userEvent.type(usernameElement, 'lili');
//   const buttonElement = screen.getByRole('button');
//   await userEvent.click(buttonElement);
//   const welcomeElement = screen.getAllByText(/lili/i);
//   expect(welcomeElement).toBeInTheDocument();
// });
