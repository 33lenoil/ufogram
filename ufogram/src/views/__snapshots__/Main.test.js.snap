// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the component matches the snapshot 1`] = `
<div
  style={
    Object {
      "display": "flex",
      "justifyContent": "space-evenly",
    }
  }
>
  <div
    style={
      Object {
        "background": "#8769b6",
        "display": "flex",
        "justifyContent": "center",
        "position": "fixed",
        "width": "100%",
      }
    }
  >
    <h1>
      UFOgram 
      lionelhu
    </h1>
    <div
      style={
        Object {
          "display": "flex",
          "flexDirection": "row",
        }
      }
    >
      <input
        name="SearchUser"
        onChange={[Function]}
        type="text"
      />
      <button
        onClick={[Function]}
        title="searchUser"
        type="button"
      >
        Search
      </button>
    </div>
    <div>
      <button
        onClick={[Function]}
        title="My Profile"
        type="button"
      >
        My Profile
      </button>
      <button
        onClick={[Function]}
        title="Create New Post"
        type="button"
      >
        Create New Post
      </button>
    </div>
  </div>
  <div
    style={
      Object {
        "display": "flex",
        "justifyContent": "center",
        "marginTop": "80px",
        "width": "100%",
      }
    }
  >
    <div
      style={
        Object {
          "background": "#b6f486",
          "display": "flex",
          "flexDirection": "column",
          "justifyContent": "center",
          "width": "100%",
        }
      }
    >
      <t>
        No posts
      </t>
    </div>
  </div>
</div>
`;
